feat: add useComponentWillReceiveUpdate (#24)
SukkaW/foxact: docs/src/pages/use-component-will-receive-update.mdx

---
title: useComponentWillReceiveUpdate
---

# useComponentWillReceiveUpdate

import ExportMetaInfo from '../components/export-meta-info';

<ExportMetaInfo />

> Reset part of states when props changed.

Reviewer: ```suggestion
This hook allows you to change the state based on information from the previous render during a re-render. The name of the hook comes from [`UNSAFE_componentWillReceiveProps` method of class components](https://react.dev/reference/react/Component#unsafe_componentwillreceiveprops).
```

## Usage

```js
import { useComponentWillReceiveUpdate } from 'foxact/use-abortable-effect';

function Component(props) {
  const [a, setA] = useState('')
  const [b, setB] = useState('')
  // when props.x changed, only reset a, not b
  useComponentWillReceiveUpdate(() => {
    setA('')
  }, [props.x]);
}
```

If you're using useEffect like this:

```js
  const [state, setState] = useState(false)
  useEffect(() => setState(false), [props.someProp])
```

Don't do it. See [Adjusting some state when a prop changes](https://react.dev/learn/you-might-not-need-an-effect#adjusting-some-state-when-a-prop-changes) or [Storing information from previous renders](https://react.dev/reference/react/useState#storing-information-from-previous-renders)
It should be like this:
```js
  const [prev, setPrev] = useState(state)
  if (prev !== state) {
    setPrev(state)
    setState(false)
  }
```

This hook is a helper for the above pattern.

```js
useComponentWillReceiveUpdate(() => setState(false), [state])
```

This only applies to states of the current component.

> Modifying states from other components causes React reporting errors.

Reviewer: Mention that this is your last resort when working with third-party libraries/components/SDKs.

You may also want to read [(Avoid) Notifying parent components about state changes](https://react.dev/learn/you-might-not-need-an-effect#notifying-parent-components-about-state-changes) and [(Avoid) Passing data to the parent](https://react.dev/learn/you-might-not-need-an-effect#passing-data-to-the-parent).
If you need to edit other components' states, write it like this:

```js
useComponentWillReceiveUpdate(() => {

>   setLocalState(false)

Reviewer: ```suggestion
import { flushSync } from 'react-dom';

useComponentWillReceiveUpdate(() => {
  // Force React to immediately flush scheduled/batched updates
  flushSync(() => setLocalState(false))
  // By the time flushSync finishes and reaches this line, the DOM update
  // for local state change has finished, you can now safely trigger parent
  // components' re-render.
```

>   Promise.resolve().then(() => props.setParentState(false))

Reviewer: Use `flushSync` from React DOM is preferred: https://react.dev/reference/react-dom/flushSync

---
title: useComponentWillReceiveUpdate
---

# useComponentWillReceiveUpdate

import ExportMetaInfo from '../components/export-meta-info';

<ExportMetaInfo />

Reset part of states when props changed.

## Usage

```js
import { useComponentWillReceiveUpdate } from 'foxact/use-abortable-effect';

function Component(props) {
  const [a, setA] = useState('')
  const [b, setB] = useState('')
  // when props.x changed, only reset a, not b
  useComponentWillReceiveUpdate(() => {
    setA('')
  }, [props.x]);
}
```

If you're using useEffect like this:

```js
  const [state, setState] = useState(false)
  useEffect(() => setState(false), [props.someProp])
```

> Don't do it. See [Adjusting some state when a prop changes](https://react.dev/learn/you-might-not-need-an-effect#adjusting-some-state-when-a-prop-changes)

Reviewer: Should also mention https://react.dev/reference/react/useState#storing-information-from-previous-renders ?

Author: those two articles look the same, I think maybe no need for attach both

Reviewer: IMHO we could link both. Also "information from previous render" definitely means more than "prop" alone.

}
```

If you're using useEffect like this:

```js
  const [state, setState] = useState(false)
  useEffect(() => setState(false), [props.someProp])
```

Don't do it. See [Adjusting some state when a prop changes](https://react.dev/learn/you-might-not-need-an-effect#adjusting-some-state-when-a-prop-changes) or [Storing information from previous renders](https://react.dev/reference/react/useState#storing-information-from-previous-renders)
It should be like this:
```js
const [prev, setPrev] = useState(state)
if (prev !== state) {
  setPrev(state)
  setState(false)
}
```

This hook is a helper for the above pattern.

```js
useComponentWillReceiveUpdate(() => setState(false), [state])
```

This only applies to states of the current component.
Modifying states from other components causes React reporting errors.
You may also want to read [(Avoid) Notifying parent components about state changes](https://react.dev/learn/you-might-not-need-an-effect#notifying-parent-components-about-state-changes) and [(Avoid) Passing data to the parent](https://react.dev/learn/you-might-not-need-an-effect#passing-data-to-the-parent).
If you need to edit other components' states, write it like this:

```js

> useComponentWillReceiveUpdate(() => {

Reviewer: ```suggestion
import { flushSync } from 'react-dom'

useComponentWillReceiveUpdate(() => {
```

## Usage

```js
import { useComponentWillReceiveUpdate } from 'foxact/use-abortable-effect';

function Component(props) {
  const [a, setA] = useState('')
  const [b, setB] = useState('')
  // when props.x changed, only reset a, not b
  useComponentWillReceiveUpdate(() => {
    setA('')
  }, [props.x]);
}
```

If you're using useEffect like this:

```js
  const [state, setState] = useState(false)
  useEffect(() => setState(false), [props.someProp])
```

Don't do it. See [Adjusting some state when a prop changes](https://react.dev/learn/you-might-not-need-an-effect#adjusting-some-state-when-a-prop-changes) or [Storing information from previous renders](https://react.dev/reference/react/useState#storing-information-from-previous-renders)
It should be like this:
```js
const [prev, setPrev] = useState(state)
if (prev !== state) {
  setPrev(state)
  setState(false)
}
```

This hook is a helper for the above pattern.

```js
useComponentWillReceiveUpdate(() => setState(false), [state])
```

> This only applies to states of the current component.
> Modifying states from other components causes React reporting errors.
> You may also want to read [(Avoid) Notifying parent components about state changes](https://react.dev/learn/you-might-not-need-an-effect#notifying-parent-components-about-state-changes) and [(Avoid) Passing data to the parent](https://react.dev/learn/you-might-not-need-an-effect#passing-data-to-the-parent).
> If you need to edit other components' states, write it like this:

Reviewer: ```suggestion
This should only apply to states of the current component. Modifying states of other components causes React reporting errors. You may also want to read [(Avoid) Notifying parent components about state changes](https://react.dev/learn/you-might-not-need-an-effect#notifying-parent-components-about-state-changes) and [(Avoid) Passing data to the parent](https://react.dev/learn/you-might-not-need-an-effect#passing-data-to-the-parent).

If you really need to directly modify other components' states (E.g. when working with third-party libraries/components/SDKs where you don't have control of that code), use `flushSync` to separate two state updates:
```
